File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    answer = []
    N = len(words)
    visited = [ False for _ in range(N)]
    def countDiff(str1, str2):
        count = 0
        for s1, s2 in zip(str1, str2):
            if(s1!=s2):count+=1
        return count
    
    def dfs(start, dept, lst):
        if(start==target):
            answer.append(dept)
            return
        for idx, v in enumerate(words):
            if(visited[idx] or countDiff(v, start)>1):
                continue
            visited[idx]=True
            dfs(v, dept+1, lst+[v])
            visited[idx]=False
    dfs(begin, 0, [begin])
    return min(answer) if len(answer)>0 else 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = [];
    const N = words.length
    const visited = Array.from({length:N}).fill(false)
    function getDiffCount(str1, str2){
        const M = str1.length
        let count = 0
        for(let idx=0;idx<M;idx++){
            if(str1[idx]!=str2[idx]){count++}
        }
        // console.log("## diff", str1, str2, count)
        return count
    }
    function dfs(dept, curr, lst){
        if(curr===target){
            // console.log(dept, lst)
            answer.push(dept)
        }
        for(let idx=0;idx<N;idx++){
            if(visited[idx]||getDiffCount(curr, words[idx])>1)continue
            visited[idx]=true
            dfs(dept+1, words[idx], [...lst, words[idx]])
            visited[idx]=false
            
        }
        
    }
    dfs(0, begin, [begin])
    // console.log("###", answer)
    return answer.length>0?Math.min(...answer):0;
}